README.md
# Pusher Channels Java Client

[![Build Status](https://travis-ci.org/pusher/pusher-websocket-java.svg?branch=master)](https://travis-ci.org/pusher/pusher-websocket-java)
[![codecov](https://codecov.io/gh/pusher/pusher-websocket-java/branch/master/graph/badge.svg)](https://codecov.io/gh/pusher/pusher-websocket-java)
[![Maven Central](https://img.shields.io/maven-central/v/com.pusher/pusher-java-client.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22com.pusher%22%20AND%20a:%22pusher-java-client%22)

Pusher Channels client library for Java targeting **Android** and general Java.

## Supported platforms

* Java SE - supports versions 8, 11 and 17
* Oracle JDK
* OpenJDK
* Android 7 and above. 5 and 6 will require [desugaring](https://developer.android.com/studio/write/java8-support#library-desugaring).

## TOC

This README covers the following topics:

- [Pusher Channels Java Client](#pusher-channels-java-client)
  - [Supported platforms](#supported-platforms)
  - [TOC](#toc)
  - [Installation](#installation)
    - [Maven](#maven)
    - [Gradle](#gradle)
    - [Download](#download)
    - [Source](#source)
  - [API Overview](#api-overview)
  - [The Pusher constructor](#the-pusher-constructor)
  - [Connecting](#connecting)
  - [The PusherOptions object](#the-pusheroptions-object)
  - [Reconnecting](#reconnecting)
  - [Disconnecting](#disconnecting)
  - [Listening to connection events](#listening-to-connection-events)
  - [Subscribing to channels](#subscribing-to-channels)
    - [Public channels](#public-channels)
    - [Private channels](#private-channels)
    - [Private encrypted channels](#private-encrypted-channels)
    - [Presence channels](#presence-channels)
      - [The User object](#the-user-object)
      - [Client event authenticity](#client-event-authenticity)
  - [Binding and handling events](#binding-and-handling-events)
    - [ChannelEventListener](#channeleventlistener)
    - [SubscriptionEventListener](#subscriptioneventlistener)
    - [Unbinding event listeners](#unbinding-event-listeners)
    - [Example](#example)
  - [Triggering events](#triggering-events)
  - [Accessing the connection socket ID](#accessing-the-connection-socket-id)
  - [Helper Methods](#helper-methods)
    - [Getting a channel from string](#getting-a-channel-from-string)
      - [Basic channels](#basic-channels)
      - [Private channels](#private-channels-1)
      - [Presence channels](#presence-channels-1)
    - [Check if a channel has subscribed](#check-if-a-channel-has-subscribed)
  - [JavaDocs](#javadocs)
  - [Library Development Environment](#library-development-environment)
    - [Prerequisites](#prerequisites)
    - [Cloning the project](#cloning-the-project)
    - [Android Studio](#android-studio)
    - [Eclipse Project](#eclipse-project)
    - [Build](#build)
    - [Run the Example Application](#run-the-example-application)

## Installation

The compiled library is available in two ways:

### Maven

The pusher-java-client is available in Maven Central.

```xml
<dependencies>
    <dependency>
      <groupId>com.pusher</groupId>
      <artifactId>pusher-java-client</artifactId>
      <version>2.4.4</version>
    </dependency>
</dependencies>
```

### Gradle

```groovy
dependencies {
  implementation 'com.pusher:pusher-java-client:2.4.4'
}
```

### Download

You can download a version of the `.jar` directly from <https://repo1.maven.org/maven2/com/pusher/pusher-java-client/>

### Source

You can build the project from the source in this repository. See **Library development environment** for more information on build environment.

## API Overview

Here's the API in a nutshell.

```java
// Create a new Pusher instance
PusherOptions options = new PusherOptions().setCluster(YOUR_APP_CLUSTER);
Pusher pusher = new Pusher(YOUR_APP_KEY, options);

pusher.connect(new ConnectionEventListener() {
    @Override
    public void onConnectionStateChange(ConnectionStateChange change) {
        System.out.println("State changed to " + change.getCurrentState() +
                           " from " + change.getPreviousState());
    }

    @Override
    public void onError(String message, String code, Exception e) {
        System.out.println("There was a problem connecting!");
    }
}, ConnectionState.ALL);

// Subscribe to a channel
Channel channel = pusher.subscribe("my-channel");

// Bind to listen for events called "my-event" sent to "my-channel"
channel.bind("my-event", new SubscriptionEventListener() {
    @Override
    public void onEvent(PusherEvent event) {
        System.out.println("Received event with data: " + event.toString());
    }
});

// Disconnect from the service
pusher.disconnect();

// Reconnect, with all channel subscriptions and event bindings automatically recreated
pusher.connect();
// The state change listener is notified when the connection has been re-established,
// the subscription to "my-channel" and binding on "my-event" still exist.
```

More information in reference format can be found below.

## The Pusher constructor

The standard constructor take an application key which you can get from the app's API Access section in the Pusher Channels dashboard.

```java
PusherOptions options = new PusherOptions().setCluster(YOUR_APP_CLUSTER);
Pusher pusher = new Pusher(YOUR_APP_KEY, options);
```

If you are going to use [private](https://pusher.com/docs/channels/using_channels/private-channels) or [presence](https://pusher.com/docs/channels/using_channels/presence-channels) channels then you will need to provide an `ChannelAuthorizer` to be used when authenticating subscriptions. In order to do this you need to pass in a `PusherOptions` object which has had an `ChannelAuthorizer` set.

```java
HttpChannelAuthorizer channelAuthorizer = new HttpChannelAuthorizer("http://example.com/some_auth_endpoint");
PusherOptions options = new PusherOptions().setCluster(YOUR_APP_CLUSTER).setChannelAuthorizer(channelAuthorizer);
Pusher pusher = new Pusher(YOUR_APP_KEY, options);
```

See the documentation on [Authorizing Users](https://pusher.com/docs/channels/server_api/authorizing-users) for more information.

If you need finer control over the endpoint then the setHost, setWsPort and setWssPort methods can be employed.
## Connecting

In order to send and receive messages you need to connect to Channels.

```java
PusherOptions options = new PusherOptions().setCluster(YOUR_APP_CLUSTER);
Pusher pusher = new Pusher(YOUR_APP_KEY, options);
pusher.connect();
```

## The PusherOptions object

Most of the functionality of this library is configured through the PusherOptions object. You configure it by calling
methods with parameters on the object before passing it to the Pusher object. Below is a table containing all of the
methods you can call.

| Method                      | Parameter         | Description                                                                                                                                   |
|-----------------------------|-------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
| setEncrypted                | Boolean           | Sets whether the connection should be made with TLS or not.                                                                                   |
| setChannelAuthorizer        | ChannelAuthorizer | Sets the channel authorizer to be used when authorizing private and presence channels.                                                        |
| setHost                     | String            | The host to which connections will be made.                                                                                                   |
| setWsPort                   | int               | The port to which unencrypted connections will be made. Automatically set correctly.                                                          |
| setWssPort                  | int               | The port to which encrypted connections will be made. Automatically set correctly.                                                            |
| setCluster                  | String            | Sets the cluster the client will connect to, thereby setting the Host and Port correctly.                                                     |
| setActivityTimeout          | long              | The number of milliseconds of inactivity at which a "ping" will be triggered to check the connection. The default value is 120,000.           |
| setPongTimeout              | long              | The number of milliseconds the client waits to receive a "pong" response from the server before disconnecting. The default value is 30,000.   |
| setMaxReconnectionAttempts  | int               | Number of reconnection attempts that will be made when pusher.connect() is called, after which the client will give up.                       |
| setMaxReconnectGapInSeconds | int               | The delay in two reconnection extends exponentially (1, 2, 4, .. seconds) This property sets the maximum inbetween two reconnection attempts. |
| setProxy                    | Proxy             | Specify a proxy, e.g. ```options.setProxy( new Proxy( Proxy.Type.HTTP, new InetSocketAddress( "proxyaddress", 80 ) ) )```                     |

## Reconnecting

The `connect` method is also used to re-connect in case the connection has been lost, for example if an Android
device loses reception. Note that the state of channel subscriptions and event bindings will be preserved while
disconnected and re-negotiated with the server once a connection is re-established.

## Disconnecting

```java
pusher.disconnect();
```

After disconnection the Pusher instance will release any internally allocated resources (threads and network connections)

## Listening to connection events

Implement the `ConnectionEventListener` interface to receive connection state change events:

```java
PusherOptions options = new PusherOptions().setCluster(YOUR_APP_CLUSTER);
Pusher pusher = new Pusher(YOUR_APP_KEY, options);
pusher.connect(new ConnectionEventListener() {
    @Override
    public void onConnectionStateChange(ConnectionStateChange change) {
        System.out.println("State changed to " + change.getCurrentState() +
                           " from " + change.getPreviousState());
    }

    @Override
    public void onError(String message, String code, Exception e) {
        System.out.println("There was a problem connecting!");
    }
});
```

A series of `ConnectionState` members can be passed after the listener in this call to filter the states which will receive notification, e.g.

```java
// MyConnectionEventListener is notified only of transitions to the disconnected state
pusher.connect(new MyConnectionEventListener(), ConnectionState.DISCONNECTED);
```

For more information see [connection states](https://pusher.com/docs/channels/using_channels/connection#connection-states).

## Subscribing to channels

Channels uses the concept of [channels](https://pusher.com/docs/channels/using_channels/channels) as a way of subscribing to data. They are identified and subscribed to by a simple name. Events are bound to on a channels and are also identified by name. To listen to an event you need to implemented the `ChannelEventListener` interface (see **Binding and handling events**).

As mentioned above, channel subscriptions need only be registered once per `Pusher` instance. They are preserved across disconnection and re-established with the server on reconnect. They should NOT be re-registered. They may, however, be registered with a `Pusher` instance before the first call to `connect` - they will be completed with the server as soon as a connection becomes available.

### Public channels

```java
Channel channel = pusher.subscribe("my-channel");
```

If you wish to be informed when the subscription succeeds, pass an implementation of the `ChannelEventListener` interface:

```java
Channel channel = pusher.subscribe("my-channel", new ChannelEventListener() {
    @Override
    public void onSubscriptionSucceeded(String channelName) {
        System.out.println("Subscribed to channel: " + channelName);
    }

    // Other ChannelEventListener methods
});
```
If you wish to be informed for subscription count events, use the `bind` function to listen to event type `pusher:subscription_count`:

```java
Channel channel = pusher.subscribe("my-channel");
channel.bind("pusher:subscription_count", new SubscriptionEventListener() {
    @Override
    public void onEvent(PusherEvent event) {
        System.out.println("Received event with data: " + event.toString());
        System.out.println("Subscription Count is: " + channel.getCount());
    }
});

```

### Private channels

It's possible to subscribe to [private channels](https://pusher.com/docs/channels/using_channels/private-channels) that provide a mechanism for [authorizing channel subscriptions](https://pusher.com/docs/channels/server_api/authorizing-users). In order to do this you need to provide a `ChannelAuthorizer` when creating the `Pusher` instance (see **The Pusher constructor** above).

The library provides a `HttpChannelAuthorizer` implementation of `ChannelAuthorizer` which makes an HTTP `POST` request to an authorization endpoint. However, you can implement your own authorization mechanism if required.

Private channels are subscribed to as follows:

```java
PrivateChannel privateChannel = pusher.subscribePrivate( "private-channel" );
```

In addition to the events that are possible on public channels a private channel exposes an `onAuthenticationFailure`. This is called if the `ChannelAuthorizer` does not successfully authorize the subscription:

```java
PrivateChannel channel = pusher.subscribePrivate("private-channel",
    new PrivateChannelEventListener() {
        @Override
        public void onAuthenticationFailure(String message, Exception e) {
            System.out.println(
                String.format("Authentication failure due to [%s], exception was [%s]", message, e)
            );
        }

        // Other ChannelEventListener methods
    });
```

### Private encrypted channels

Similar to Private channels, you can also subscribe to a
[private encrypted channel](https://pusher.com/docs/channels/using_channels/encrypted-channels).
This library now fully supports end-to-end encryption. This means that only you and your connected clients will be able to read your messages. Pusher cannot decrypt them.

Like the private channel, you must provide your own authorization endpoint,
with your own encryption master key. There is a
[demonstration endpoint to look at using nodejs](https://github.com/pusher/pusher-channels-auth-example#using-e2e-encryption).

To get started you need to subscribe to your channel, provide a `PrivateEncryptedChannelEventListener`, and a list of the events you are
interested in, for example:

```java
PrivateEncryptedChannel privateEncryptedChannel =
	pusher.subscribePrivateEncrypted("private-encrypted-channel", listener, "my-event");
```

In addition to the events that are possible on public channels the
`PrivateEncryptedChannelEventListener` also has the following methods:
* `onAuthenticationFailure(String message, Exception e)` - This is called if
the `ChannelAuthorizer` does not successfully authorize the subscription:
* `onDecryptionFailure(String event, String reason);` - This is called if the message cannot be
decrypted. The decryption will attempt to refresh the shared secret key once
from the `ChannelAuthorizer`.

There is a
[working example in the repo](https://github.com/pusher/pusher-websocket-java/blob/master/src/main/java/com/pusher/client/example/PrivateEncryptedChannelExampleApp.java)
which you can use with the
[demonstration authorization endpoint](https://github.com/pusher/pusher-channels-auth-example#using-e2e-encryption)

### Presence channels

[Presence channels](https://pusher.com/docs/channels/using_channels/presence-channels) are private channels which provide additional events exposing who is currently subscribed to the channel. Since they extend private channels they also need to be authorized (see [authorizing channel subscriptions](https://pusher.com/docs/channels/server_api/authorizing-users)).

Presence channels can be subscribed to as follows:

```java
PresenceChannel presenceChannel = pusher.subscribePresence( "presence-channel" );
```

Presence channels provide additional events relating to users joining (subscribing) and leaving (unsubscribing) the presence channel. It is possible to listen to these events by implementing the `PresenceChannelEventListener`.

```java
PresenceChannel channel = pusher.subscribePresence("presence-channel",
    new PresenceChannelEventListener() {
        @Override
        public void onUsersInformationReceived(String channelName, Set<User> users) {
            for (User user : users) {
                userSubscribed(channelName, user);
            }
        }

        @Override
        public void userSubscribed(String channelName, User user) {
            System.out.println(
                String.format("A new user joined channel [%s]: %s, %s",
                              channelName, user.getId(), user.getInfo())
            );

            if (user.equals(channel.getMe())) {
                System.out.println("me");
            }
        }

        @Override
        public void userUnsubscribed(String channelName, User user) {
            System.out.println(
                String.format("A user left channel [%s]: %s %s",
                              channelName, user.getId(), user.getInfo())
            );
        }

        // Other ChannelEventListener methods
    });
```

#### The User object

*Note: In the [Pusher Channels documentation](http://pusher.com/docs/channels) a User may be referred to as a `Member`.*

The `User` object has two main methods.

`getId` fetches a unique identifier for the user on the presence channel.

`getInfo` fetches a string representing arbitrary additional information about the user in the form of a JSON hash, e.g.

```
{"user_name":"Mr. User","user_score":1357}
```

The following example using the [Gson library](https://sites.google.com/site/gson/gson-user-guide) to handle deserialization:

```java
String jsonInfo = user.getInfo();
Gson gson = new Gson();
UserInfo info = gson.fromJson(jsonInfo, UserInfo.class);
```

For more information on defining the user id and user info on the server see [Implementing the authorization endpoint for a presence channel](https://pusher.com/docs/channels/server_api/authorizing-users#implementing-the-authorization-endpoint-for-a-presence-channel) documentation.

#### Client event authenticity

Channels now provides a 'user-id' with client events sent from the server. With presence channels, your authorization endpoint provides your user with a user-id. Previously, it was up to you to include this user-id in every client-event triggered by your clients. Now, when a client of yours triggers a client event, Channels will append their user-id to their triggered message, so that the other clients in the channel receive it. This allows you to trust that a given user really did trigger a given payload.

If you’d like to make use of this feature, you’ll need to extract the user-id from the message delivered by Channels. To do this, call getUserId() on the event payload your event handler gets called with, like so:

```java
channel.bind("client-my-event", new SubscriptionEventListener() {
    @Override
    public void onEvent(PusherEvent event) {
        System.out.println("Received event with userId: " + event.getUserId());
    }
});
```

## Binding and handling events

There are two types of events that occur on channel subscriptions.

1. Protocol related events such as those triggered when a subscription succeeds
2. Application events that have been triggered by code within your application

### ChannelEventListener


CHANGELOG.md
# Changelog

## 2.4.4

- [CHANGED] pusher.pro narrower proguard rules.

## 2.4.3

- [FIXED] Fix issue with json serialization when using proguard

## Version 2.4.2 - 23th Sep 2022
* Fixes a crash when an event comes in without and escaped data json member.
* update to com.google.code.gson:gson:2.9.1
* update to org.java-websocket:Java-WebSocket:1.5.3

## Version 2.4.1 - 13th Sep 2022
* Refactoring and code cleanup of event handling in the SDK
* Fixes subscription_count events

## Version 2.4.0 - 15th July 2022
* Add support for Subscription count events

## Version 2.3.0 - 4th July 2022
* Added support for user sign in and server to user messages
* Fixed issue with calling disconnect while the client is attempting reconnection
* Fixed issue with calling connect while the client is disconnecting

## Version 2.2.8 - 30th November 2021
* Add global [event listner](https://github.com/pusher/pusher-websocket-java/issues/305)

## Version 2.2.7 - 11th November 2021
* Fixed an [issue Vulnerability with legacy Gson dependency](https://github.com/pusher/pusher-websocket-java/issues/305)

## Version 2.2.6 - 17th March 2021
* Fixed an [issue causing a NullPointerException if event data on a Presence channel was malformed or missing](https://github.com/pusher/pusher-websocket-java/pull/290)

## Version 2.2.5 - 11th August 2020
* Fixed an [issue with the Java-WebSocket library with android pre-sdk24 devices causing a NoSuchMethod exception](https://github.com/pusher/pusher-websocket-java/pull/284)

## Version 2.2.4 - 3rd August 2020
* Removed unnecessary code in the TweetNaClFast implementation
* Updated the Java-WebSocket library to 1.5.1 as the [current version 1.4.0 has a known critical vulnerability.](https://github.com/pusher/pusher-websocket-java/issues/279)

## Version 2.2.3 - 8th July 2020

* Fixed LazySodium causing a crash on Android devices by swapping it out for [TweetNaCl](https://github.com/InstantWebP2P/tweetnacl-java)

## Version 2.2.2 - 6th July 2020

* Fixed an issue where some private encrypted messages were not decrypted accurately by swapping our implementation of SecretBoxOpener for one provided by [Lazy Sodium](https://github.com/terl/lazysodium-java)

## Version 2.2.1 - 22nd April 2020

* Changed PusherOptions `setForceTLS` and `isForceTLS` to `setUseTLS` and `isUseTLS` to align with the other client SDKs.

## Version 2.2.0 - 22nd April 2020

* Changed PusherOptions `setEncrypted` and `isEncrypted` to `setForceTLS` and `isForceTLS` to reduce confusion between this option and private encrypted channels.

## Version 2.1.1 - 15th April 2020

* Fix a case where multiple websocket connections could be opened at once
  if reconnection was triggered by an activity timeout.

## Version 2.1.0 - 8th April 2020

* Added support for [private encrypted channels](https://pusher.com/docs/channels/using_channels/encrypted-channels)

## Version 2.0.2

* Make a no-op state change non-fatal, allowing transitions to/from the same state

## Version 2.0.1

* Forward `channel_data` returned by the auth endpoint when subscribing to private channels

## Version 2.0.0

* The onEvent handler now gets called with one parameter, called the PusherEvent. This PusherEvent has all the same information available as was available before with the 3 parameters, but now is accessible by calling getChannelName(), getData() or getEventName() on the PusherEvent object you receive. In addition, for presence channel client events, you can now retrieve an authenticated User ID by calling getUserId() on the PusherEvent object. To read more on Authenticated users, see the README or our docs [here](https://pusher.com/docs/channels/using_channels/events#user-id-in-client-events).
* Update Java sourceCompatibility and targetCompatibility to 1.8.
* Fix an issue where the reconnect logic would not be reset after connect() is called again.
* Depend on `org.java-websocket:Java-WebSocket:1.4.0` instead of `com.pusher:java-websocket:1.4.1`.

## Version 1.4.0
* Update the dependency to use pusher/java-websocket fork and remove dependency on clojars.org repository.

## Version 1.3.0
* Add retry logic when the connection is lost
* Accept 201 status code from auth endpoints

## Version 1.2.2

 * Improve resillience of pong timeout disconnections

## Version 1.2.0

2016-07-05 jpatel531

  * Expose Client interface

## Version 1.1.3

2016-05-12 mdpye, plackemacher, jpatel531

  * Fix exceptions where tasks were being pushed onto a shutting-down event queue.
  * Reduce construction of GSON objects and keep one centralized instance.

## Version 1.1.2

2016-03-23 jpatel531

  * Fix cases where the library would throw an error due to a connection state change
    from `CONNECTING` to `CONNECTING`. More information [here](https://github.com/pusher/pusher-websocket-java/pull/102)
  * Fix `ConcurrentModificationException` on event listeners.

## Version 1.1.1

2016-03-22 jpatel531

  * Removes the dependency on slf4j-log4j

## Version 1.1.0

2016-03-09 jpatel531, jameshfisher

  * Allow specifying a proxy via which to connect to Pusher.

## Version 1.0.2

2015-11-06 leggetter, siggijons
  * Use @SerializedName in PresenceChannelImpl for better serialization support across languages e.g. Turkish

## Version 1.0.1

2015-11-06 hamchapman, jpatel531
  * Resolves issues where Gson would cast numeric user ids as doubles before converting them to a string, leading to inconsistencies

## Version 1.0.0

2015-10-7
  * Use generic `Map` for `HttpAuthorizer` - trevorrjohn
  * Add to `Pusher` `isSubscribed` `getChannel`, `getPrivateChannel` and `getPresenceChannel` methods - jpatel531
  * Library unsubscribes asynchronously - jpatel531
  * Synchronize access to channel event listeners - trevorrjohn, jpatel531, mdpye

## Version 0.3.3
2015-02-09 mdpye
 * REALLY remove the JavaWebsockets submodule

## Version 0.3.2

src/main/java/com/pusher/client/PusherOptions.java
package com.pusher.client;

import com.pusher.pusher_websocket_java.BuildConfig;

import java.net.Proxy;

/**
 * Configuration for a {@link com.pusher.client.Pusher} instance.
 */
public class PusherOptions {

    public static final String LIB_VERSION = BuildConfig.APP_VERSION;

    private static final String URI_SUFFIX = "?client=java-client&protocol=5&version=" + LIB_VERSION;
    private static final String WS_SCHEME = "ws";
    private static final String WSS_SCHEME = "wss";

    private static final int WS_PORT = 80;
    private static final int WSS_PORT = 443;
    private static final String PUSHER_DOMAIN = "pusher.com";

    private static final long DEFAULT_ACTIVITY_TIMEOUT = 120000;
    private static final long DEFAULT_PONG_TIMEOUT = 30000;

    private static final int MAX_RECONNECTION_ATTEMPTS = 6; //Taken from the Swift lib
    private static final int MAX_RECONNECT_GAP_IN_SECONDS = 30;

    // Note that the primary cluster lives on a different domain
    // (others are subdomains of pusher.com). This is not an oversight.
    // Legacy reasons.
    private String host = "ws.pusherapp.com";
    private int wsPort = WS_PORT;
    private int wssPort = WSS_PORT;
    private boolean useTLS = true;
    private long activityTimeout = DEFAULT_ACTIVITY_TIMEOUT;
    private long pongTimeout = DEFAULT_PONG_TIMEOUT;
    private UserAuthenticator userAuthenticator;
    private ChannelAuthorizer channelAuthorizer;
    private Authorizer authorizer;
    private Proxy proxy = Proxy.NO_PROXY;
    private int maxReconnectionAttempts = MAX_RECONNECTION_ATTEMPTS;
    private int maxReconnectGapInSeconds = MAX_RECONNECT_GAP_IN_SECONDS;

    /**
     * @deprecated Please use isUseTLS
     */
    @Deprecated
    public boolean isEncrypted() {
        return useTLS;
    }

    /**
     * @deprecated Please use setUseTLS
     */
    @Deprecated
    public PusherOptions setEncrypted(final boolean encrypted) {
        this.useTLS = encrypted;
        return this;
    }

    /**
     * @return whether the connection to Pusher should use TLS
     */
    public boolean isUseTLS() {
        return useTLS;
    }

    /**
     * Sets whether the connection to Pusher should be use TLS.
     *
     * @param useTLS whether the connection should use TLS, by default this is true
     * @return this, for chaining
     */
    public PusherOptions setUseTLS(final boolean useTLS) {
        this.useTLS = useTLS;
        return this;
    }

    /**
     * Gets the user authenticator to be used when signing in.
     *
     * @return the user authenticator
     */
    public UserAuthenticator getUserAuthenticator() {
        return userAuthenticator;
    }

    /**
     * Sets the user authenticator to be used when signing in.
     *
     * @param userAuthenticator The user authenticator to be used.
     * @return this, for chaining
     */
    public PusherOptions setUserAuthenticator(final UserAuthenticator userAuthenticator) {
        this.userAuthenticator = userAuthenticator;
        return this;
    }

    /**
     * Gets the channel authorizer to be used when authorizing private and presence
     * channels.
     *
     * @return the channel authorizer
     */
    public ChannelAuthorizer getChannelAuthorizer() {
        return channelAuthorizer;
    }

    /**
     * Sets the channel authorizer to be used when authorizing private and presence
     * channels.
     *
     * @param channelAuthorizer The channel authorizer to be used.
     * @return this, for chaining
     */
    public PusherOptions setChannelAuthorizer(final ChannelAuthorizer channelAuthorizer) {
        this.channelAuthorizer = channelAuthorizer;
        return this;
    }

    /**
     * @deprecated Please use getChannelauthorizer
     */
    @Deprecated
    public Authorizer getAuthorizer() {
        return authorizer;
    }

    /**
     * @deprecated Please use setChannelauthorizer
     */
    @Deprecated
    public PusherOptions setAuthorizer(final Authorizer authorizer) {
        this.authorizer = authorizer;
        return setChannelAuthorizer(authorizer);
    }

    /**
     * The host to which connections will be made.
     * <p>
     * Note that if you wish to connect to a standard Pusher cluster, the
     * convenience method setCluster will set the host and ports correctly from
     * a single argument.
     *
     * @param host The host
     * @return this, for chaining
     */
    public PusherOptions setHost(final String host) {
        this.host = host;
        return this;
    }

    /**
     * The port to which non TLS connections will be made.
     * <p>
     * Note that if you wish to connect to a standard Pusher cluster, the
     * convenience method setCluster will set the host and ports correctly from
     * a single argument.
     *
     * @param wsPort port number
     * @return this, for chaining
     */
    public PusherOptions setWsPort(final int wsPort) {
        this.wsPort = wsPort;
        return this;
    }

    /**
     * The port to which encrypted connections will be made.
     * <p>
     * Note that if you wish to connect to a standard Pusher cluster, the
     * convenience method setCluster will set the host and ports correctly from
     * a single argument.
     *
     * @param wssPort port number
     * @return this, for chaining
     */
    public PusherOptions setWssPort(final int wssPort) {
        this.wssPort = wssPort;
        return this;
    }

    public PusherOptions setCluster(final String cluster) {
        host = "ws-" + cluster + "." + PUSHER_DOMAIN;
        wsPort = WS_PORT;
        wssPort = WSS_PORT;
        return this;
    }

    /**
     * The number of milliseconds of inactivity at which a "ping" will be
     * triggered to check the connection.
     * <p>
     * The default value is 120,000 (2 minutes). On some connections, where
     * intermediate hops between the application and Pusher are aggressively
     * culling connections they consider to be idle, a lower value may help
     * preserve the connection.
     *
     * @param activityTimeout time to consider connection idle, in milliseconds
     * @return this, for chaining
     */
    public PusherOptions setActivityTimeout(final long activityTimeout) {
        if (activityTimeout < 1000) {
            throw new IllegalArgumentException(
                    "Activity timeout must be at least 1,000ms (and is recommended to be much higher)"
            );
        }

        this.activityTimeout = activityTimeout;
        return this;
    }

    public long getActivityTimeout() {
        return activityTimeout;
    }

    /**
     * The number of milliseconds after a "ping" is sent that the client will
     * wait to receive a "pong" response from the server before considering the
     * connection broken and triggering a transition to the disconnected state.
     * <p>
     * The default value is 30,000.
     *
     * @param pongTimeout time to wait for pong response, in milliseconds
     * @return this, for chaining
     */
    public PusherOptions setPongTimeout(final long pongTimeout) {
        if (pongTimeout < 1000) {
            throw new IllegalArgumentException("Pong timeout must be at least 1,000ms (and is recommended to be much higher)");
        }

        this.pongTimeout = pongTimeout;
        return this;
    }

    /**
     * Number of reconnect attempts when websocket connection failed
     *
     * @param maxReconnectionAttempts number of max reconnection attempts, default = {@link #MAX_RECONNECTION_ATTEMPTS} 6
     * @return this, for chaining
     */
    public PusherOptions setMaxReconnectionAttempts(int maxReconnectionAttempts) {
        this.maxReconnectionAttempts = maxReconnectionAttempts;
        return this;
    }

    /**
     * The delay in two reconnection extends exponentially (1, 2, 4, .. seconds) This property sets the maximum in between two
     * reconnection attempts.
     *
     * @param maxReconnectGapInSeconds time in seconds of the maximum gab between two reconnection attempts, default = {@link #MAX_RECONNECT_GAP_IN_SECONDS} 30s
     * @return this, for chaining
     */
    public PusherOptions setMaxReconnectGapInSeconds(int maxReconnectGapInSeconds) {
        this.maxReconnectGapInSeconds = maxReconnectGapInSeconds;
        return this;
    }

    public long getPongTimeout() {
        return pongTimeout;
    }

    /**
     * Construct the URL for the WebSocket connection based on the options
     * previous set on this object and the provided API key
     *
     * @param apiKey The API key
     * @return the WebSocket URL
     */
    public String buildUrl(final String apiKey) {
        return String.format(
                "%s://%s:%s/app/%s%s",
                useTLS ? WSS_SCHEME : WS_SCHEME,
                host,
                useTLS ? wssPort : wsPort,
                apiKey,
                URI_SUFFIX
        );
    }

    /**
     * The default value is Proxy.NO_PROXY.
     *
     * @param proxy Specify a proxy, e.g. <code>options.setProxy( new Proxy( Proxy.Type.HTTP, new InetSocketAddress( "proxyaddress", 80 ) ) )</code>;
     * @return this, for chaining
     */
    public PusherOptions setProxy(Proxy proxy) {
        if (proxy == null) {
            throw new IllegalArgumentException("proxy must not be null (instead use Proxy.NO_PROXY)");
        }
        this.proxy = proxy;
        return this;
    }

    /**
     * @return The proxy to be used when opening a websocket connection to Pusher.
     */
    public Proxy getProxy() {
        return this.proxy;
    }

    /**
     * @return the maximum reconnection attempts
     */
    public int getMaxReconnectionAttempts() {
        return maxReconnectionAttempts;
    }

    /**
     * @return the maximum reconnection gap in seconds
     */
    public int getMaxReconnectGapInSeconds() {
        return maxReconnectGapInSeconds;
    }
}
